### PR TITLE
Implement retries strategy

### PR DESCRIPTION
**Pull request by michalsidzej**
Resolves L2B-2235

### CODE REVIEW

File: packages/uif/src/BaseIndexer.test.ts [hunk 1]
```diff
@@ -55,15 +59,106 @@ describe(BaseIndexer.name, () => {
...
 
       await parent.doTick(5)
+      await parent.finishTick(5)
 
       expect(middle.getState().waiting).toEqual(true)
     })
   })
+
+  describe('retries on error', () => {
+    // Those tests are quite slow (around 500ms total)
```
Comment (sz-piotr) [inline packages/uif/src/BaseIndexer.test.ts:75]: Make them faster

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/uif/src/BaseIndexer.test.ts [hunk 2]
```diff
@@ -55,15 +59,106 @@ describe(BaseIndexer.name, () => {
...
+
+      await parent.doTick(1)
+      await parent.finishTick(1)
+
+      await child.finishUpdate(new Error('test error'))
+      expect(child.updating).toBeFalsy()
+      expect(child.getState().status).toEqual('invalidating')
+
+      await child.finishInvalidate()
+      await clock.tickAsync(1000)
```
Comment (sz-piotr) [inline packages/uif/src/BaseIndexer.test.ts:103]: add idle check here and in other tests too

File: packages/uif/src/BaseIndexer.ts [hunk 3]
```diff
@@ -14,6 +14,13 @@ import {
   UpdateEffect,
 } from './reducer/types/IndexerEffect'
 import { IndexerState } from './reducer/types/IndexerState'
+import { Retries, RetryStrategy } from './Retries'
+
+const DEFAULT_RETRY_STRATEGY = Retries.exponentialBackOff({
```
Comment (sz-piotr) [inline packages/uif/src/BaseIndexer.ts:19]: Make this configurable, maybe a static field in BaseIndexer

File: packages/uif/src/Retries.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+function exponentialBackOff(opts: ExponentialBackOffOpts): RetryStrategy {
+  let attempts = 0
+  const maxAttempts = opts.maxAttempts
+  assert(maxAttempts > 0)
+  const maxDistanceMs = opts.maxDistanceMs ?? Infinity
+  assert(maxDistanceMs > 0)
+
+  return {
+    shouldRetry: () => {
+      attempts++
```
Comment (sz-piotr) [inline packages/uif/src/Retries.ts:31]: Maybe this function should be indempotent? And we can have another function `markAttempt`?

File: packages/uif/src/Retries.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      return attempts <= maxAttempts
+    },
+    timeoutMs: () => {
+      const distance = Math.pow(2, attempts) * opts.stepMs
```
Comment (sz-piotr) [inline packages/uif/src/Retries.ts:35]: Do we start with `2 * opts.stepMs`? Why not `1 * opts.stepMs`

File: packages/uif/src/Retries.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import assert from 'node:assert'
+
+export interface RetryStrategy {
+  shouldRetry: () => boolean
+  timeoutMs: () => number
+  clear: () => void
+}
+
+interface ExponentialBackOffOpts {
+  stepMs: number
```
Comment (sz-piotr) [inline packages/uif/src/Retries.ts:10]:
```suggestion
  initialTimeoutMs: number
```

File: packages/uif/src/Retries.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  maxAttempts: number
+  maxDistanceMs?: number
```
Comment (sz-piotr) [inline packages/uif/src/Retries.ts:12]:
```suggestion
  maxTimeoutMs?: number
```

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/uif/src/BaseIndexer.test.ts (hunk 1 above)
Comment (michalsidzej) [reply packages/uif/src/BaseIndexer.test.ts]: c1ddd5b

File: packages/uif/src/BaseIndexer.test.ts (hunk 2 above)
Comment (michalsidzej) [reply packages/uif/src/BaseIndexer.test.ts:112]: fe935f3

File: packages/uif/src/BaseIndexer.ts (hunk 3 above)
Comment (michalsidzej) [reply packages/uif/src/BaseIndexer.ts]: 5f2b23a

File: packages/uif/src/Retries.ts (hunk 4 above)
Comment (michalsidzej) [reply packages/uif/src/Retries.ts]: 5f2b23a

**Review by sdlyy — APPROVED** [review]

**Pull request merged**